### PR TITLE
Introduce FC HBA statistics collection+reporting

### PR DESCRIPTION
**Pull request by stevekay**
Collect HBA statistics when -S DISK used with sadc.
Report HBA statistics when -a used with sar.

### CODE REVIEW

File: rd_stats.c [hunk 1]
```diff
@@ -2104,5 +2104,59 @@ void read_filesystem(struct stats_filesystem *st_filesystem, int nbr)
...
+	/* Each HBA, if present, will have its own hostX entry within SYSFS_HBA */
+	if ((dir = opendir(SYSFS_HBA)) == NULL)
+		return; /* No HBAs */
+	
+	while ((drd = readdir(dir)) != NULL) {
+		if(strncmp(drd->d_name,"host",4)==0) {
+			sprintf(word_stats,"%s/%s/statistics/tx_frames",SYSFS_HBA,drd->d_name);
+			if ((fp = fopen(word_stats, "r"))) {
+				if(fgets(line,sizeof(line),fp))	
+					sscanf(line, "%x",&tx_frames);
```
Comment (sysstat) [inline rd_stats.c]: I'm wondering why you read them with %x. Is it an hexadecimal number?

File: rd_stats.c [hunk 2]
```diff
@@ -2104,5 +2104,59 @@ void read_filesystem(struct stats_filesystem *st_filesystem, int nbr)
...
+				fclose(fp);
+			}
+
+			sprintf(word_stats,"%s/%s/statistics/rx_frames",SYSFS_HBA,drd->d_name);
+			if ((fp = fopen(word_stats, "r"))) {
+				if(fgets(line,sizeof(line),fp))	
+					sscanf(line, "%x",&rx_frames);
```
Comment (sysstat) [inline rd_stats.c]: Ditto

File: activity.c [hunk 3]
```diff
@@ -1226,6 +1226,37 @@ struct activity filesystem_act = {
...
+#endif
+#ifdef SOURCE_SAR
+	.f_print	= print_hba_stats,
+	.f_print_avg	= print_avg_hba_stats,
+#endif
+#ifdef SOURCE_SADF
+	.f_render	= render_hba_stats,
+	.f_xml_print	= xml_print_hba_stats,
+	.f_json_print	= json_print_hba_stats,
+	.hdr_line	= "HBA;HBAtxframes;HBArxframes",
```
Comment (sysstat) [inline activity.c]: Please use the same metrics names than those displayed by sar (ie. "txframes/s" and not "HBAtxframes")

File: json_stats.c [hunk 4]
```diff
@@ -2163,3 +2163,43 @@ __print_funct_t json_print_filesystem_stats(struct activity *a, int curr, int ta
...
+	for (i = 0; i < a->nr; i++) {
+		sfc = (struct stats_hba *) ((char *) a->buf[curr]  + i * a->msize);
+
+		if (sep) {
+			printf(",\n");
+		}
+		sep = TRUE;
+
+		xprintf0(tab, "{\"hba\": \"%s\", "
+			 "\"txframes\": %.0f, "
```
Comment (sysstat) [inline json_stats.c]: Ditto

File: man/sar.in [hunk 5]
```diff
@@ -166,7 +166,39 @@ command only reports on local activities.
...
+
+The following values are displayed:
+
+.B txframes
+.RS
+.RS
+Total number of frames transmitted.
+.RE
+
+.B txframes/s
```
Comment (sysstat) [inline man/sar.in]: Ditto

File: pr_stats.c [hunk 6]
```diff
@@ -2582,3 +2582,75 @@ __print_funct_t print_avg_filesystem_stats(struct activity *a, int prev, int cur
...
+ */
+__print_funct_t stub_print_hba_stats(struct activity *a, int prev, int curr, int dispavg, unsigned long long itv)
+{
+	int i;
+	struct stats_hba *sfc,*sfp;
+
+
+	if (dis) {
+		printf("\n%-11s        HBA         txframe  txframe/s         rxframe  rxframe/s\n",
+		       (dispavg ? _("Summary:") : timestamp[curr]));
```
Comment (sysstat) [inline pr_stats.c]: This won't be a Summary but an Average once absolute frames numbers will be removed.

File: rd_stats.h [hunk 7]
```diff
@@ -547,6 +547,15 @@ struct stats_filesystem {
 
 #define STATS_FILESYSTEM_SIZE	(sizeof(struct stats_filesystem))
 
+/* Structure for HBA statistics */
+struct stats_hba {
+	unsigned long long f_txframes		__attribute__ ((aligned (16)));
+	unsigned long long f_rxframes		__attribute__ ((aligned (16)));
```
Comment (sysstat) [inline rd_stats.h]: Defined as unsigned long long here.
Read as unsigned int in rd_stats.c.
Please be consistent (use the same type as used by the kernel. If you don't feel browsing the kernel sources, then use unsigned long long).

File: rndr_stats.c [hunk 8]
```diff
@@ -2917,3 +2917,41 @@ __print_funct_t render_filesystem_stats(struct activity *a, int isdb, char *pre,
...
+					int curr, unsigned long long itv)
+{
+	int i;
+	struct stats_hba *sfc;
+
+	for (i = 0; i < a->nr; i++) {
+		sfc = (struct stats_hba *) ((char *) a->buf[curr] + i * a->msize);
+
+		render(isdb, pre, PT_USERND ,
+		       "%s\tHBA",
```
Comment (sysstat) [inline rndr_stats.c]: txframes should appear somewhere. I don't see it...?

**Comment by sysstat** [comment]
Steve,
I'm reviewing your pull request for HBA statistics. Please could you:
* Send me the contents of a /sys/class/fc_host directory tree (ls -lR /sys/class/fc_host)
* Send me the contents of rx_frames and tx_frames files located in this tree (cat /sys/class/.../{rx,tx}_frames)
* Clean your pull request by removing all the intermediate/debug commits (eg. README is modified then reverted to original, debug messages are inserted then removed, sar manual page is modified several times...) On the other hand, keep several small patches as you have done, this is good.

Other remarks:
* Shall we really talk about HBA statistics or Fibre Channel ones? May files in /sys/class/fc_host contain other statistics than Fibre Channel ones? Moreover, I would rather use something like "-n HBA" (or "-n FC") to display those statistics instead of using option -a (we will soon be short of new possible options with sar ;-) )
* I'm also wondering why we would need to display absolute frames number? This is not the logic we have for other activities like block devices or network interfaces. Please remove absolute numbers and keep only per-second ones.
* Please also have a look at the various comments I made on your code.
Thanks.

File: rd_stats.c (hunk 1 above)
Comment (jaunix) [inline rd_stats.c]: Yes.  Comes through as hex.  Example below.  

    # cat /sys/class/fc_host/host12/statistics/tx_frames
    0x2e32

Exhaustive info about typical outputs at https://community.emc.com/servlet/JiveServlet/previewBody/12212-102-1-46887/Linux_Fabric_Enquiry_Native_Commands_CLI_v1.2.pdf

**Comment by jaunix** [comment]
Thanks for the various suggestions.
As it sounds like there's an appetite for this sort of functionality, I'll give the the code an overhaul in the next few days, incorporating all those points raised, and resubmit.

**Comment by stevekay** [comment]
Closing this pull request, will resubmit an overhauled version in a few days.

**Pull request closed without merging**